### PR TITLE
rust: Use extension policy mechanism to check for unaccounted critical extensions

### PR DESCRIPTION
**Pull request by tetsuo-cpp**

### CODE REVIEW

File: src/rust/cryptography-x509-validation/src/policy/mod.rs [hunk 1]
```diff
@@ -369,6 +356,21 @@ impl<'a, B: CryptoOps> Policy<'a, B> {
             }
         }
 
+        // Check that all critical extensions in this certificate are accounted for.
+        for extension in extensions.iter() {
```
Comment (tetsuo-cpp) [inline src/rust/cryptography-x509-validation/src/policy/mod.rs]: Wasn't exactly sure how best to unit test this in Rust. Perhaps via the chain validation API (so in `cryptography-x509-validation/src/lib.rs`)?

File: src/rust/cryptography-x509-validation/src/policy/mod.rs [hunk 2]
```diff
@@ -369,6 +356,21 @@ impl<'a, B: CryptoOps> Policy<'a, B> {
...
+            if extension.critical {
+                let policies_handle_extension = |policies: &Vec<ExtensionPolicy<B>>| {
+                    policies.iter().any(|x| x.oid == extension.extn_id)
+                };
+                if !policies_handle_extension(&self.common_extension_policies)
+                    && !policies_handle_extension(&self.ca_extension_policies)
+                    && !policies_handle_extension(&self.ee_extension_policies)
+                {
+                    return Err("certificate contains unaccounted critical extension".into());
+                }
```
Comment (woodruffw) [inline src/rust/cryptography-x509-validation/src/policy/mod.rs]: FWIW, you can probably reduce this indirection by doing `Iterator::chain` + `Iteratory::any` instead, or set subtraction between the two. I'll play around with it and push something up 🙂

File: src/rust/cryptography-x509-validation/src/policy/mod.rs (hunk 1 above)
Comment (woodruffw) [reply src/rust/cryptography-x509-validation/src/policy/mod.rs]: I _think_ we should probably do it with a limbo test (if there isn't one already?) -- a leaf/intermediate with an unknown critical extension should fail validation.

(It's unclear to me whether a root with an unknown critical extension should fail validation -- roots are defined as a representation of a trust authority, and a TA is defined just in terms of a public key and subject.)

File: src/rust/cryptography-x509-validation/src/policy/mod.rs [hunk 3]
```diff
@@ -369,6 +356,28 @@ impl<'a, B: CryptoOps> Policy<'a, B> {
...
+            .map(|p| p.oid.clone())
+            .collect::<HashSet<_>>();
+        let unchecked_extensions = critical_extensions
+            .difference(&checked_extensions)
+            .collect::<Vec<_>>();
+
+        if !unchecked_extensions.is_empty() {
+            // TODO: Render the OIDs here.
+            return Err("certificate contains unaccounted-for critical extensions".into());
+        }
```
Comment (woodruffw) [inline src/rust/cryptography-x509-validation/src/policy/mod.rs:379]: This is slightly more verbose, but should be more efficient than the previous critical extension sweep.

**Pull request merged**